### PR TITLE
Haar 1220 create email domain tests

### PR DESCRIPTION
**Pull request by olatomoloju**
Please help out!

### CODE REVIEW

File: backend/controllers/emailDomainValidation.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 const isValidDomainNameChars = (domainName: string) => domainName.match(/^[0-9A-Za-z-,.()'&]*$/)
-const isValidDomainDescriptionChars = (domainDesc: string) => domainDesc.match(/^[0-9A-Z_]*$/)
+const isValidDomainDescriptionChars = (domainDesc: string) => domainDesc.match(/^[0-9A-Z_ ]*$/)
```
Comment (ravmoj) [inline backend/controllers/emailDomainValidation.ts:2]: Is additional space required ? [0-9A-Z_ ]

File: integration-tests/e2e/emaildomain.create.cy.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  it('Should display error and prevent form submission, when email domain name is blank', () => {
+    cy.task('stubSignIn', { roles: [{ roleCode: 'MAINTAIN_EMAIL_DOMAINS' }] })
+    cy.signIn()
+    cy.task('stubAllEmailDomains', {})
+    MenuPage.verifyOnPage().viewEmailDomainListing()
+    const emailDomainListingPage = EmailDomainListingPage.verifyOnPage()
+    emailDomainListingPage.navigateToCreateEmailDomainPageButton().click()
+    const emailDomainCreationPage = EmailDomainCreationPage.verifyOnPage()
+    emailDomainCreationPage.createEmailDomainPage('', 'Domain1Description')
```
Comment (ravmoj) [inline integration-tests/e2e/emaildomain.create.cy.js:48]: is the space related to this test?

File: backend/controllers/emailDomainValidation.ts (hunk 1 above)
Comment (olatomoloju) [reply backend/controllers/emailDomainValidation.ts]: Yes, I have updated the regex to /^[0-9A-Z_\s+]*$/

File: integration-tests/e2e/emaildomain.create.cy.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    })
+  })
+
+  it('Should display error and prevent form submission, when email domain description is blank', () => {
+    cy.task('stubSignIn', { roles: [{ roleCode: 'MAINTAIN_EMAIL_DOMAINS' }] })
+    cy.signIn()
+    cy.task('stubAllEmailDomains', {})
+    MenuPage.verifyOnPage().viewEmailDomainListing()
+    const emailDomainListingPage = EmailDomainListingPage.verifyOnPage()
+    emailDomainListingPage.navigateToCreateEmailDomainPageButton().click()
```
Comment (karenmillermoj) [inline integration-tests/e2e/emaildomain.create.cy.js:34]: Probably easier/neater to just have a method that gets you to the create email domain page - then all the tests can use it

File: backend/controllers/addEmailDomain.test.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        flash: jest.fn(),
+        originalUrl: '/email-domains',
+      }
+      await emailDomainFactory.post(req, { locals, redirect })
+      expect(redirect).toBeCalledWith('/email-domains')
+      expect(req.flash).toBeCalledWith('createEmailDomainErrors', [
+        { href: '#domainName', text: 'Domain name already exists' },
+      ])
+    })
+  })
```
Comment (karenmillermoj) [inline backend/controllers/addEmailDomain.test.js:124]: are we missing a CSRF token test?

File: integration-tests/e2e/emaildomain.create.cy.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    const emailDomainListingPage = EmailDomainListingPage.verifyOnPage()
+    emailDomainListingPage.navigateToCreateEmailDomainPageButton().click()
+
+    const emailDomainCreationPage = EmailDomainCreationPage.verifyOnPage()
+    emailDomainCreationPage.cancel()
+    emailDomainListingPage.emailDomainListingTableRows().should('have.length', 3)
+    emailDomainListingPage.emailDomainListingTableRows().eq(0).should('include.text', 'DOMAIN1')
+    emailDomainListingPage.emailDomainListingTableRows().eq(0).should('include.text', 'DOMAIN 1 DESCRIPTION')
+  })
```
Comment (karenmillermoj) [inline integration-tests/e2e/emaildomain.create.cy.js:80]: are we missing a CSRF token check?

File: integration-tests/e2e/emaildomain.create.cy.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    cy.task('stubSignIn', { roles: [{ roleCode: 'MAINTAIN_EMAIL_DOMAINS' }] })
+    cy.signIn()
+    cy.task('stubAllEmailDomains', {})
+    MenuPage.verifyOnPage().viewEmailDomainListing()
+    const emailDomainListingPage = EmailDomainListingPage.verifyOnPage()
+    emailDomainListingPage.navigateToCreateEmailDomainPageButton().click()
+    const emailDomainCreationPage = EmailDomainCreationPage.verifyOnPage()
+    emailDomainCreationPage.createEmailDomainPage('Domain1', 'Domain1Description')
+    cy.task('verifyGetAllEmailDomains').should((requests) => {
+      expect(requests).to.have.lengthOf(1)
```
Comment (karenmillermoj) [inline integration-tests/e2e/emaildomain.create.cy.js:24]: Not sure what this test is doing - the verifyGetAllEmailDomains just verifies that the method has been called

File: integration-tests/e2e/emaildomain.create.cy.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    const emailDomainListingPage = EmailDomainListingPage.verifyOnPage()
+    emailDomainListingPage.navigateToCreateEmailDomainPageButton().click()
+    const emailDomainCreationPage = EmailDomainCreationPage.verifyOnPage()
+    emailDomainCreationPage.createEmailDomainPage('', '')
+    emailDomainCreationPage.errorSummary().should('contain.text', 'Enter a domain description')
+    emailDomainCreationPage.errorSummary().should('contain.text', 'Enter a domain name')
+  })
+
+  it('Should navigate to email domain listing, on-clicking the  cancel create email domain button', () => {
```
Comment (karenmillermoj) [inline integration-tests/e2e/emaildomain.create.cy.js:66]: ` the  can ` extra spacing - file maybe needs formatting?

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 2 above)
Comment (olatomoloju) [reply integration-tests/e2e/emaildomain.create.cy.js:48]: Yes, the domain desc allows spaces but not the domain name.

File: backend/controllers/addEmailDomain.test.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      const req = { flash: jest.fn() }
+      const flashDomain = req.flash('domain')
+      const domain = flashDomain != null && flashDomain.length > 0 ? flashDomain[0] : ''
+      const render = jest.fn()
+      await emailDomainFactory.index(standardReq, { render })
+      expect(render).toBeCalledWith('createEmailDomain.njk', {
+        createEmailDomainUrl: '/create-email-domain',
+        listEmailDomainUrl: '/email-domains',
+        ...domain,
+        errors: req.flash('createEmailDomainErrors'),
```
Comment (karenmillermoj) [inline backend/controllers/addEmailDomain.test.js:21]: Shouldn't the errors be undefined at this point?

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 3 above)
Comment (olatomoloju) [reply integration-tests/e2e/emaildomain.create.cy.js:34]: Probably, yes.

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 5 above)
Comment (olatomoloju) [reply integration-tests/e2e/emaildomain.create.cy.js:80]: CSRF Defo. 
The disallowing of direct access when user does not have role is already covered in the email domain listing which is the entry point, but can replicate for create email domain.

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 6 above)
Comment (olatomoloju) [reply integration-tests/e2e/emaildomain.create.cy.js:24]: Yes, it does just that. The create email  domain calls the email domain listing, so its a good check.

File: backend/controllers/emailDomainValidation.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
 const isValidDomainNameChars = (domainName: string) => domainName.match(/^[0-9A-Za-z-,.()'&]*$/)
-const isValidDomainDescriptionChars = (domainDesc: string) => domainDesc.match(/^[0-9A-Z_]*$/)
+const isValidDomainDescriptionChars = (domainDesc: string) => domainDesc.match(/^[0-9A-Z_\s+]*$/)
```
Comment (karenmillermoj) [inline backend/controllers/emailDomainValidation.ts:2]: we need tests for emailDomainValidation.ts

File: backend/controllers/addEmailDomain.test.js (hunk 8 above)
Comment (olatomoloju) [reply backend/controllers/addEmailDomain.test.js:21]: Yes, req.flash('xxx') is undefined and can put undefined for clarity.

**Comment by olatomoloju** [comment]
All done, Thanks both.

File: backend/controllers/emailDomainValidation.ts (hunk 9 above)
Comment (karenmillermoj) [inline backend/controllers/emailDomainValidation.ts:2]: I'm not convinced these are correct - do we really allow (), etc in email domains but not the description?  seems odd.  This is possibly a bug from previous work so may need a separate PR to investigate

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 6 above)
Comment (karenmillermoj) [reply integration-tests/e2e/emaildomain.create.cy.js:40]: I don't think this is still quite right.  We are still only checking that the getAllDomainsList has been called - not what the test specifies.

File: backend/controllers/emailDomainValidation.ts (hunk 9 above)
Comment (karenmillermoj) [reply backend/controllers/emailDomainValidation.ts:2]: Is this new validation regex ?

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 6 above)
Comment (karenmillermoj) [reply integration-tests/e2e/emaildomain.create.cy.js:40]: I don't think this is still quite right. We are still only checking that the getAllDomainsList has been called - not what the test specifies. If you're doing the test using the verify - it should verifyAddEmailDomain method is called. Are you stubbing this method out?

File: integration-tests/e2e/emaildomain.create.cy.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    MenuPage.verifyOnPage()
+    cy.get('[data-qa="view_email_domains_link"]').should('not.exist')
+    cy.visit('/create-email-domain', { failOnStatusCode: false })
+    AuthErrorPage.verifyOnPage()
+  })
+
+  it('Should create new email domain, after completing form and clicking the create email domain button', () => {
+    const emailDomainCreationPage = navigateToCreateEmailDomainPage()
+    emailDomainCreationPage.createEmailDomainPage('Domain1', 'Domain1Description')
+    cy.task('verifyGetAllEmailDomains').should((requests) => {
```
Comment (karenmillermoj) [inline integration-tests/e2e/emaildomain.create.cy.js:39]: I don't think this is still quite right. We are still only checking that the getAllDomainsList has been called - not what the test specifies. If you're doing the test using the verify - it should verifyAddEmailDomain method is called. Are you stubbing this method out?

File: backend/controllers/emailDomainValidation.ts (hunk 9 above)
Comment (olatomoloju) [reply backend/controllers/emailDomainValidation.ts:2]: That is out of scope of this ticket. But, its a quick regex update when the requirement is confirmed. Thomas out BA can be involved in that.

File: integration-tests/e2e/emaildomain.create.cy.js (hunk 10 above)
Comment (olatomoloju) [reply integration-tests/e2e/emaildomain.create.cy.js:39]: Please, have a look at the code and ping me for a hurdle. We can be having long chatty conversation here! Its easier.

Comment (olatomoloju) [reply integration-tests/e2e/emaildomain.create.cy.js:39]: @karenmillermoj - Ping me on slack when you have a minute!

**Comment by olatomoloju** [comment]
All done as per screen conversation.

File: backend/controllers/emailDomainValidation.ts (hunk 9 above)
Comment (olatomoloju) [reply backend/controllers/emailDomainValidation.ts]: It was suggested validation should be removed to match existing basic empty string and string length check.

**Review by karenmillermoj — DISMISSED** [review]
lgtm

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**